types/math/dec.go

```diff
@@ -0,0 +1,126 @@
+package math
+
+import (
+	"github.com/cockroachdb/apd/v2"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+type Dec struct {
```

Comment:
Why do you say we're trying to use them as unsigned? I think we only are restricting that they're unsigned in `SafeSub`, which will be used sometimes when dealing with token balances (and wanting to ensure a non-negative result when subtracting two numbers).  But more generally speaking, I don't see why we would want to restrict this struct to only work with unsigned decimals.  There could be valid reasons for us storing ecological measurements as negative numbers (e.g. temperature) at some point down the line.

Reply:
Just updated with documentation about the wrapper (quoting from @aaronc in the original PR description).

Reply:
> Why do you say we're trying to use them as unsigned? I think we only are restricting that they're unsigned in `SafeSub`, which will be used sometimes when dealing with token balances (and wanting to ensure a non-negative result when subtracting two numbers). But more generally speaking, I don't see why we would want to restrict this struct to only work with unsigned decimals. There could be valid reasons for us storing ecological measurements as negative numbers (e.g. temperature) at some point down the line.

Okay, well then I think we're doing some mixing of concepts here. I don't think the decimal type should be holding information about the fact that it will be used as a non-negative token balance. I.e. I don't think an `InsufficientFunds` error should be thrown from within the API of this type. I also then don't think the name `SafeSub` is quite accurate, because it's not unsafe to subtract and get a negative result.

I guess there's a few options here:
- Have a separate type for explicitly unsigned decimals, for cases such as token balances where they'll never be negative
- Rename `SafeSub` to something like `SubNonNegative` (not catchy)
- Don't expose a `SafeSub` function and do the checking and error throwing elsewhere

Does that make sense?

Reply:
Personally I'm in favour of the first option. That way we're explicit in the type system about the kind of values that we're representing.

Reply:
I think we could also then have a type alias for token balances that returns the `ErrInsufficientFunds` error in it's own `SafeSub` method.

```
type Balance = UDec

func (*Balance) SafeSub ...
```

Reply:
So I opted for moving this just out of Dec struct, and just migrated to a function in the math package with signature:
```go
func SubBalance(x Dec, y Dec) (Dec, error)
```

Seems like @aaronc also had desire for a SafeAdd type operation, so I've also implemented an `AddBalance()` similar to the `SubBalance()` above.

I do think we should probably think about refactoring these, but probably better to do when we're actually looking at how these functions are used in context (presumably when actually performing arithmetic on ecocredits). So I'd like to merge this in, and when we get to making use of the `Dec` struct for tracking ecocredit balances, we then take another look at where the "safe" arithmetic options should be.

Also holding that from @aaronc's comments today, it seems that we're trying to align the implementation here with a yet-to-be completed refactor of `sdk.Dec` (note: sdk.Dec does not have an implementation of `SafeSub()`, but [sdk.DecCoin does](https://github.com/cosmos/cosmos-sdk/blob/master/types/dec_coin.go#L312-L317)).  I don't think it makes sense to have a `DecCoin` struct in this package (as ecocredits are not necessarily coins...), but maybe it makes sense to have some struct with more explicit coin-like behavior (e.g. treating it more like an unsigned dec) when refactoring the actual balance tracking to use the Dec?

Reply:
Yeah, I think at some point an explicitly unsigned Dec type would be good. Then we can property test it properly and know that we have some nice safety guarantees on it.

Reply:
I prefer `Safe*` - this is what we have in other blockchain and it's more familiar to developers.